tests/e2e-evm/test/callcode.test.ts

```diff
@@ -0,0 +1,247 @@
+              }),
+            ],
+          }),
+        }),
+        // msg.sender is the caller contract
+        wantSender: (ctx) => ctx.lowLevelCaller.address,
+        // msg.value is still 100 since the value is passed to callcode()
+        wantValue: 100n,
+      },
+      {
+        name: "callcode with storage",
+        txParams: (ctx) => ({
+          to: ctx.lowLevelCaller.address,
+          value: 0n,
+          gas: contractCallerGas + 20_000n,
+          data: encodeFunctionData({
+            abi: lowLevelCaller.abi,
+            functionName: "functionCallCode",
+            args: [
+              ctx.implementationContract.address,
+              encodeFunctionData({
+                abi: ctx.implementationContract.abi,
+                functionName: "setStorageValue",
+                args: [2n],
+              }),
+            ],
+          }),
+        }),
+        // Storage in caller contract
+        wantStorageContract: (ctx) => ctx.lowLevelCaller.address,
+        wantStorageValue: 2n,
+      },
+    ];
+
+    for (const tc of testCases) {
+      it(tc.name, async function () {
+        const txData = tc.txParams(ctx);
+        // Signer is the whale
+        txData.account = whaleAddress;
+
+        if (!txData.to) {
+          expect.fail("to field not set");
+        }
+
+        const startingBalance = await publicClient.getBalance({ address: txData.to });
+
+        const res = await publicClient.call(txData);
+
+        // Check the return value for the msg.sender and msg.value if applicable
+        if (tc.wantSender) {
+          if (!res.data) {
+            // Fail this way as a type guard to ensure res.data is not undefined
+            expect.fail("no data returned");
+          }
+
+          // Expect all test cases call the getMsgInfo function
+          const [returnAddress, returnValue] = decodeFunctionResult({
+            abi: ctx.implementationContract.abi,
+            functionName: "getMsgInfo",
+            data: res.data,
+          });
+
+          const expectedSender = tc.wantSender(ctx, whaleAddress);
+
+          // getAddress to ensure both are checksum encoded
+          expect(getAddress(returnAddress)).to.equal(getAddress(expectedSender), "unexpected msg.sender");
+          expect(returnValue).to.equal(tc.wantValue, "unexpected msg.value");
+        } else {
+          expect(res.data).to.be.undefined;
+        }
+
+        const txHash = await walletClient.sendTransaction(txData);
+        const txReceipt = await publicClient.waitForTransactionReceipt({ hash: txHash });
+        expect(txReceipt.status).to.equal("success");
+
+        if (txData.gas) {
+          expect(txReceipt.gasUsed < txData.gas, "gas to not be exhausted").to.be.true;
+        }
+
+        // Storage tests if applicable
+        if (tc.wantStorageContract) {
+          const storageContract = tc.wantStorageContract(ctx);
+          const storageValue = await publicClient.getStorageAt({
```

Comment:
Good use of the storage API when the calling contract doesn't have a function to retrieve the value via call